content/languages/python/authoring/index.md

```diff
@@ -0,0 +1,252 @@
+- Stacktraces of a crashing user solution can reveal details that should not be visible,
+- Use of an assertion not suitable for the given case may lead to incorrect test results,
+- Incorrectly used assertions may produce confusing or unhelpful messages.
+
+To avoid the above problems, calls to assertion functions should respect the following rules:
+- The expected value should be calculated _before_ invoking an assertion. The `expected` parameter passed to the assertion should not be a function call expression, but a value calculated directly beforehand.
+- Appropriate assertion functions should be used for each given test. `assert_equals` is not suitable in all situations. Use `assert_approx_equals` for floating point comparisons, `expect` for tests on boolean values, `expect_error` to test error handling.
+- Some additional attention should be paid to the order of parameters passed to assertion functions. It differs between various assertion libraries, and it happens to be quite often confused by authors, mixing up `actual` and `expected` in assertion messages. For the Python testing framework, the order is `(actual, expected)`.
+- One somewhat distinctive feature of Python assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but the execution of the current test case was not stopped and following assertions continue to refer to it. This behavior can be overridden by passing the `allow_raise=True` argument to the assertion functions that support it.
+- To avoid unexpected crashes in tests, it's recommended to perform some additional assertions before assuming that the answer returned by the user solution has some particular type, form, or value. For example, if the test suite sorts the returned list to verify its correctness, an explicit assertion should be added to check whether the returned object is actually a list, and not, for example, `None`.
+
+
+## Additional restrictions
+
+### Accessing the solution file
+
+Some kata (for example, code-golf challenges or anti-cheat tests) would like to access and read the user solution file as text. It's possible and can be done by reading a file located at `/workspace/solution.txt`.
+
+### Blocking modules
+
+_TBD_
+
+
+## Example test suite
+
+Below you can find an example test suite that covers most of the common scenarios mentioned in this article. Note that it does not present all possible techniques, so actual test suites can use a different structure, as long as they keep to established conventions and do not violate authoring guidelines.
+
+
+```python
+
+#import modules explicitly
+import codewars_test as test
+import preloaded
+from solution import user_solution
+
+@test.describe('Fixed tests')
+def fixed_tests():
```

Comment:
```suggestion
def _():
```

function names for the blocks are actually useless and redundant with the block message. I generally use `_` instead. Up to you to see if you wanna change them all or none.

Reply:
I'd personally stick to meaningful names, even if they are redundant. I'd be afraid that in larger test suite it will be confusing when you'd end up with functions called `_`, `__`, `___` etc.

Reply:
It's not a problem to name them all `_` lol, you don't have to go `__` and `___` actually.

Reply:
So you telling me I can have a test suite which goes like:

```python
@test.describe('Fixed tests')
def _():
    some it blocks here...

@test.describe('Random tests')
def _():
    some it blocks here...
```

Reply:
Yes. The function name gets overridden ofc, but python doesn't care, and the decorator picks it up in time.

Reply:
WTF what kind of witchcraft is even that? :D It's definitely not something I would expect to happen.

Reply:
Lol, Python is a mutable language! XD

Reply:
Okay, it might be just me, but if I were to guess, I would not expect this, ummm... feature, to not cause problems some time in the future. If anything, if avoiding redundancy would be a goal, I'd make decorators parameterless, and derive the title of the section from decorated function name:

```python
@test.describe
def fixed_tests():
    some it blocks here...

@test.describe
def random_tests():
    some it blocks here...
```

I think I'd document such possibility in the framework reference, in its part on `it` and `describe`, and definitely would not give it as an example ;)

Putting two identical names in the same scope, _willingly_, is asking for trouble - change my mind.

@kazk any insight from technical side, if/how it could affect the runner?

Reply:
Python's decorator is a function that takes a function. `@test.describe`/`@test.it` calls the given function. See https://github.com/codewars/python-test-framework/blob/a07c83f0b5a0692f665ee316865510867ab9a4cf/codewars_test/test_framework.py#L103-L121

So it's equivalent to:
```javascript
describe('Fixed Tests', function _() {
});
describe('Random Tests', function _() {
});
```

> Putting two identical names in the same scope, 

I don't think it does that, but I was never interested in Python enough to know the details, so I can be wrong.

Reply:
I guess it does define functions `_`. But I don't think it'll do anything bad because overwriting it doesn't change the function passed to the decorator.

Reply:
```python
# Simplified decorator
def factory(output_cmd): # DESCRIBE/IT
    def decorator(name):
        def wrapper(func):
            print(output_cmd, name)
            func()
            print('completedin')
        return wrapper
    return decorator

describe = _factory('DESCRIBE')
@describe('Fixed Tests')
def _():
    # ...
# is a syntactic suger of
def _():
    # ...
_ = _factory('DESCRIBE')('Fixed Tests')(_)
```

So `_` has the return value of `wrapper(func)` which is usually `None`.

Reply:
>  I'd make decorators parameterless, and derive the title of the section from decorated function name:

I'd say it would be a bad idea, because doing so, you limit the feedback to valid functions identifiers. Then you'll need to go back to parameters to allow a well formed sentence as feedback, using paremeters again (they did that for JUnit, I beleive. It's just a mess... XD)

Reply:
I added a paragraph on this.